[WIP] Low dependency example

mohe2015 opened the pull request:
Hello,

I don't know if you want to keep this open but I tried to reduce the dependencies of this (in light of recent supply chain attacks) and maybe this is of use for somebody else. This could definitely not be merged as is as it requires manual work currently to install the package (you need to know the include path for nodejs). Also it doesn't use prebuilt binaries any more obviously and I don't intend to change that.

(Just ignore the commits, the diff is the most relevant)

Kind regards
Moritz

ranisalt commented:
Well, considering how much this hurts usability I think this goes the wrong way.

One of our runtime dependencies, `node-addon-api` is a must so we can't drop that, but it comes from Node.js itself and contains no dependencies. `@phc/format` also pulls no dependencies but is a contender for being interned because we do not require 100% of it and could live with a subset of functionalities. `@mapbox/node-pre-gyp` pulls the most dependencies and I would of course like to not depend on it.

Maybe a better approach would be to prune dependencies down the tree, in popular packages such as prettier and mocha. It really bugs me that our dev dependencies pull babel even though we do not need it at all, for example.

Interesting how this led to a substantial increase in yarn.lock packages though 🤔 try `yarn upgrade` to see if it helps.

ranisalt commented:
> That was the main one that I didn't like and removed. Another possibility would be to switch to node-gyp and not provide pre-built binaries which I tried before and which worked quite well but still has quite a few dependencies.

In the past we only used node-gyp and changed specifically because of the prebuild binaries 😆 it may be a possibility to split the library between argon2 and argon2-prebuilt or something like that.

> I just removed the yarn.lock and added package-lock.json because that's what I'm using so I'm not sure if you mislooked or tried using yarn locally.

Indeed, my bad. Why though? 🤔

mohe2015 commented:
> > I just removed the yarn.lock and added package-lock.json because that's what I'm using so I'm not sure if you mislooked or tried using yarn locally.
> 
> Indeed, my bad. Why though? thinking

Just because this is a draft and I'm not using yarn so I couldn't update yarn.lock anyways.  Another thing that shows why this is a draft.
